Neos.Neos/Classes/Fusion/NodeUriImplementation.php

```diff
@@ -160,32 +160,39 @@ public function evaluate()
         NodeAggregateId::fromString(\mb_substr($node, 7))
         );*/
 
-        $uriBuilder = new UriBuilder();
         $possibleRequest = $this->runtime->fusionGlobals->get('request');
         if ($possibleRequest instanceof ActionRequest) {
-            $uriBuilder->setRequest($possibleRequest);
+            $nodeUriBuilder = $this->nodeUriBuilderFactory->forRequest($possibleRequest->getHttpRequest());
         } else {
             // unfortunately, the uri-builder always needs a request at hand and cannot build uris without
             // even, if the default param merging would not be required
             // this will improve with a reformed uri building:
             // https://github.com/neos/flow-development-collection/pull/2744
-            $uriBuilder->setRequest(
-                ActionRequest::fromHttpRequest(ServerRequest::fromGlobals())
-            );
+            // todo correct?
+            $nodeUriBuilder = $this->nodeUriBuilderFactory->forRequest(ServerRequest::fromGlobals());
         }
-        $uriBuilder
-            ->setFormat($this->getFormat())
-            ->setCreateAbsoluteUri($this->isAbsolute())
-            ->setArguments($this->getAdditionalParams())
-            ->setSection($this->getSection())
-            ->setAddQueryString($this->getAddQueryString())
-            ->setArgumentsToBeExcludedFromQueryString($this->getArgumentsToBeExcludedFromQueryString());
-
+        // todo what to do with ???
+        //  todo getArgumentsToBeExcludedFromQueryString
+        //  todo getAddQueryString
         try {
-            return (string)NodeUriBuilder::fromUriBuilder($uriBuilder)->uriFor($nodeAddress);
+            $resolvedUri = $nodeUriBuilder->uriFor(
+                NodeAddress::fromNode($node),
+                Options::create(
+                    forceAbsolute: $this->isAbsolute(),
+                    format: $this->getFormat(),
+                    routingArguments: $this->getAdditionalParams()
+                )
```

Comment:
nitpick: large try/catch blocks always make me a bit suspicious even though it's not a real problem here.
What do you think of a construct like:
```php
$options =  Options::create(
    forceAbsolute: $this->isAbsolute(),
    format: $this->getFormat(),
    routingArguments: $this->getAdditionalParams()
 );
try {
    $resolvedUri = $nodeUriBuilder->uriFor(NodeAddress::fromNode($node), $options));
} catch (NoMatchingRouteException) {
    // ...
}
```